src/super_gradients/training/datasets/collate_fn.py

```diff
@@ -0,0 +1,421 @@
+from data_gradients.dataset_adapters.detection_adapter import DetectionDatasetAdapter
+from data_gradients.dataset_adapters.segmentation_adapter import SegmentationDatasetAdapter
+from data_gradients.dataset_adapters.classification_adapter import ClassificationDatasetAdapter
+
+from super_gradients.common.registry.registry import register_collate_function
+from super_gradients.common.decorators.factory_decorator import resolve_param
+from super_gradients.common.factories.collate_functions_factory import CollateFunctionsFactory
+from super_gradients.training.utils.detection_utils import xyxy2cxcywh
+
+from super_gradients.common.environment.device_utils import device_config
+from super_gradients.common.environment.ddp_utils import get_local_rank
+from super_gradients.training.utils.distributed_training_utils import wait_for_the_master
+from super_gradients.training.datasets.segmentation_datasets.segmentation_dataset import SegmentationDataSet
+
+
+class DatasetItemsException(Exception):
+    def __init__(self, data_sample: Tuple, collate_type: Type, expected_item_names: Tuple):
+        """
+        :param data_sample: item(s) returned by a dataset
+        :param collate_type: type of the collate that caused the exception
+        :param expected_item_names: tuple of names of items that are expected by the collate to be returned from the dataset
+        """
+        collate_type_name = collate_type.__name__
+        num_sample_items = len(data_sample) if isinstance(data_sample, tuple) else 1
+        error_msg = f"`{collate_type_name}` only supports Datasets that return a tuple {expected_item_names}, but got a tuple of len={num_sample_items}"
+        super().__init__(error_msg)
+
+
+class BaseDatasetAdapterCollateFN(ABC):
+    """Base Collate function that adapts an input dataset to SuperGradients format
+
+    This is done by applying the adapter logic either before or after the original collate function,
+    depending on whether the adapter was calibrated on a batch or a sample.
+
+    Note that the original collate function (if any) will still be used, but will be wrapped into this class.
+    """
+
+    @resolve_param("collate_fn", CollateFunctionsFactory())
+    def __init__(self, adapter: BaseDatasetAdapter, collate_fn: Optional[Callable] = None):
+        """
+        :param collate_fn:          Collate function to wrap. If None, the default collate function will be used.
+        :param adapter:             Dataset adapter to use
+        """
+        self._adapter = adapter
+        self._collate_fn = collate_fn or default_collate
+        self._adapt_on_batch = adapter.data_config.is_batch or adapter.data_config.is_batch is None
+
+        if isinstance(self._collate_fn, type(self)):
+            raise RuntimeError(f"You just tried to instantiate {self.__class__.__name__} with a `collate_fn` of the same type, which is not supported.")
+
+        self._require_calibration = not self._adapter.data_config.is_completely_initialized
```

Comment:
I don't get notion of 'calibration' in this context. Since we are using calibration in the context of quantization I feed this may confuse users (as it confuses me). 
What does "calibration" mean? Is there a better term to use?

Reply:
I agree, just kind of hoped that the person spotting this would have a better naming idea 🤣

Reply:
maybe `setup_adapter` and `require_setup` could actually do the job

Reply:
I now use `maybe_setup_dataloader_adapter` and `_maybe_setup_adapter` (which are functions)